File: day02/src/main/cpp/solution.cpp
#include <iostream>
#include <utility>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

#define VPSLL vector<pair<Direction, long long>>

using namespace std;

enum Direction {
    up,
    down,
    forward
};

Direction fromString(string value) {
    if (value == "up")
        return up;
    if (value == "down")
        return down;
    return Direction::forward;
}

pair<Direction, long long> parseLine(string line) {
    stringstream ss(line);
    string direction;
    string amount;
    ss >> direction >> amount;
    return pair(fromString(direction), stoll(amount));
}

VPSLL readInputFile() {
    ifstream file("../resources/input");
    VPSLL lines;
    string s;
    while (getline(file, s))
    {
        lines.push_back(parseLine(s));
    }
    return lines;
}

long long executeSimpleRoute(VPSLL route) {
    long long depth = 0;
    long long position = 0;
    for (auto const &value : route)
    {
        switch (value.first)
        {
        case up:
            depth -= value.second;
            break;
        case down:
            depth += value.second;
            break;
        default:
            position += value.second;
            break;
        }
    }
    return depth * position;
}

long long executeManualRoute(VPSLL route) {
    long long depth = 0;
    long long position = 0;
    long long aim = 0;
    for (auto const &value : route) {
        switch (value.first) {
        case up:
            aim -= value.second;
            break;
        case down:
            aim += value.second;
            break;
        default:
            position += value.second;
            depth += value.second * aim;
            break;
        }
    }
    return depth * position;
}

int main() {
    VPSLL route = readInputFile();
    cout << "part 1" << endl
         << executeSimpleRoute(route) << endl
         << "part 2" << endl
         << executeManualRoute(route) << endl;
    return 0;
}

File: day02/src/main/python/solution.py
def read_input():
    f = open('../resources/input', 'r')
    return list(map(lambda x: x.split(), f.readlines()))


def execute_simple_route(route):
    depth = forward = 0
    for x in route:
        direction = int(x[1])
        if x[0] == 'forward':
            forward += direction
        elif x[0] == 'down':
            depth += direction
        else:
            depth -= direction
    return depth * forward


def execute_manual_route(route):
    depth = forward = aim = 0
    for x in route:
        direction = int(x[1])
        if x[0] == 'forward':
            forward += direction
            depth += aim * direction
        elif x[0] == 'down':
            aim += direction
        else:
            aim -= direction
    return depth * forward


if __name__ == "__main__":
    route = read_input()
    print("part 1")
    print(execute_simple_route(route))
    print("part 2")
    print(execute_manual_route(route))
